src/domain/safe/entities/__tests__/erc20-transfer.factory.ts

```diff
@@ -11,6 +11,7 @@ export default function (
   value?: string,
 ): ERC20Transfer {
   return <ERC20Transfer>{
+    type: 'ERC20_TRANSFER',
```

Comment:
I believe this is only part of the list endpoint and not returned when a a single transfer is requested.

Even though the types match we should keep this in mind.

Reply:
Oh, I see, so do you think removing this hardcoded value from the factory and passing the type as an argument would be enough?

Reply:
Maybe this is something that we need to align on. I think these factories can be seen as "Domain" factories and maybe it's better to see them as "json payload" factories because that's how we have been using them.

So in this example `type: 'ERC20_TRANSFER'` would be part of the JSON payload but would not necessarily be part of the Domain object that results from that.

Not sure if this was super clear 😅

Reply:
Yes, I get your point, these factories are being used to reproduce the responses gathered from external services (transaction service, config service, currency conversion APIs, etc.), but until now we don't have any separation between our _domain objects_ and these responses. I mean, we were designing our _domain object_ shapes just to represent the payload coming from these external services, and I thought it was intentional and for the sake of simplicity.

I'm not against doing a separation between the _JSON payload_ coming from external services and our domain layer entities, but as you said that requires an alignment. 

In the current architecture, we are doing mappings/transformations between _domain objects_ and _route objects_ in the services classes, which belong to the routes layer. In the past, I commented that maybe that transformation/mapping could be done in the domain layer, and the mapping between domain objects and route objects could be trivial, just 1:1 mappings. That could resolve the problem of not having a 1:1 mapping between _JSON payloads_ and _domain objects_. (And also the -in the case of this application- minor design problem like having transformation logic coupled to the HTTP routes layer).

Another option to solve this _problem_ is to maintain the current _domain-to-route object_ transformations and add another transformation layer between JSON payloads and the domain. But honestly, I think this is a bit too much (maintaining two transformation layers) for this kind of application.

Of course this deserves its own discussion if you think it's better to change the design. IMHO we could just follow the current architecture at least until the service v1 is finally released 😅

Reply:
> In the current architecture, we are doing mappings/transformations between domain objects and route objects in the services classes, which belong to the routes layer. In the past, I commented that maybe that transformation/mapping could be done in the domain layer, and the mapping between domain objects and route objects could be trivial, just 1:1 mappings.

The decision to do this was mostly because route entities are rarely shared between routes. So you have the domain which corresponds to the Transaction Service and Config Service domains and then each route maps that domain to a specific entity. Or do you see that the route entities are mostly shared and therefore it makes sense for them to be our domain instead of the current one?

> Of course this deserves its own discussion if you think it's better to change the design. IMHO we could just follow the current architecture at least until the service v1 is finally released

I'm not sure if I understand the design change that you mention. We have:

- Domain entities
- Factories for those domain entities

Factories are used in the tests to mock the return of the HTTP layer. However if AJV serializes data in a different way (eg.: date serialisation to `Date` instead of `String`) then the factory might not match 1:1 the domain entity and that's ok! So the factory would have the date as a `String` but the resulting domain entity would have it as `Date` because it was then parsed by AJV.

Reply:
Ok, so if I understand correctly, and based on the discussion we had today, this factory should return a plain object instead of a domain object, but should also maintain `type: 'ERC20_TRANSFER'` in the payload, because that field is in the JSON payload coming from the transaction service 🙂

Reply:
> Ok, so if I understand correctly, and based on the discussion we had today, this factory should return a plain object instead of a domain object, but should also maintain `type: 'ERC20_TRANSFER'` in the payload, because that field is in the JSON payload coming from the transaction service 🙂

Yes exactly 👍 AJV would use that field to know which schema to use for validation!

Reply:
Changed [here]([39c06d7](https://github.com/5afe/safe-client-gateway-nest/pull/206/commits/39c06d7085e75619177e496b3664338ad4a550c3))